(Upstream) replace global variables, some legacy text rendering fixes

poire-z opened the pull request:
`(Upstream) Allow faster LVDocView::getBookmark()`
`(Upstream) Fix scroll mode at start and end of book`
`(Upstream) Fix sentence selection in TTS`
`(Upstream) Fix issues with legacy text rendering`
`(Upstream) Fix FB2 footnotes in legacy text rendering`
`(Upstream) Some CRC32 management fixes`
`(Upstream) FB3/DocX/ODT: get lang and description metadata`
`(Upstream) HyphMan: tweak handling of obsolete pdb files`
`(Upstream) remove global gHangingPunctuationEnabled`
`(Upstream) remove global gRenderBlockRenderingFlags`
`(Upstream) remove global gDOMVersionRequested`
`(Upstream) remove global gInterlineScaleFactor`
Commits from: https://github.com/buggins/coolreader/pull/156 https://github.com/buggins/coolreader/pull/157 https://github.com/buggins/coolreader/pull/160 https://github.com/buggins/coolreader/pull/171 https://github.com/buggins/coolreader/pull/173
Nothing much noticable for KOReader - that's just to avoid diverting too much from upstream for code we update often on our side (so, easying their future syncs with us).

`Minor fixes after global variables replacements`
- consistent default for hanging punctuation: false https://github.com/buggins/coolreader/pull/171#discussion_r497530260
- initialize _DOMVersionRequested in constructor
- use gDOMVersionCurrent instead of 0 in case some   code is used when there is no document
- bump cache file version (not really necessary, but it avoids wondering about the "style hash mismatch" that will happen on all previously opened books because the global variables values are now in different hashes).

`Add macros BLOCK_RENDERING_D/N()`
For shorter lines of code and better readability https://github.com/buggins/coolreader/pull/171#discussion_r495482151

`Fix FB2 footnotes alignment in legacy rendering`
Proposed at https://github.com/buggins/coolreader/pull/157#issuecomment-683778100 but not picked. I think it makes sense.

`lvstsheet.cpp: fix some clang-tidy warnings`
Fix some warning, discussed below.

Pinging @virxkane for info and review.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/crengine/384)
<!-- Reviewable:end -->

poire-z commented:
After the added checks for `_doc != NULL` in lvstsheep.cpp, clang-tidy complain we don't do it in some other places.
I think it makes no sense not having _doc in LVStyleSheet::parse() as we need a doc to get element name ids from. So, I'll probably just add an early check:
```diff
 bool LVStyleSheet::parse( const char * str, bool higher_importance, lString16 codeBase )
 {
+    if ( !_doc ) {
+        // We can't parse anything if no _doc to get element name Ids from
+        return false;
+    }
```
and remove all the others.

Frenzie reviewed (approved):
Didn't look at upstream much, the rest lgtm

poire-z commented:
OK, added a commit (too tired to squash it properly into previous commits) where I removed a few checks for `doc != NULL` (so that clang-tidy don't complain we don't also check in other places) where it can't happen or it make no sense to work without.

virxkane commented:
@poire-z Sorry, but I'm busy so far, I'll see it later, in a couple of days.

poire-z commented:
(Merging, as it needs some other tweaks in base and frontend, that I don't want to postpone - comments still welcome.)

virxkane commented:
@poire-z About commit ``Fix FB2 footnotes alignment in legacy rendering``:
From https://github.com/buggins/coolreader/pull/157 :

> You have:
> 
> ```css
> title p, h1 p, h2 p { 
>     $title.all
> }
> ```
> 
> where we have:
> 
> ```css
> title p, subtitle p, h1 p, h2 p, h3 p, h4 p, h5 p, h6 p {
>         text-align: center;
>         text-indent: 0px
> }
> ```
> 
> So, I guess you have that configurable in the UI. And may be if you can set titles centered there, you'll get a rendering like I do.

This is not the only reason, if you restore the block in fb2.css
```css
body[name="comments"] section title p {
    display: inline; /* counteract default of block, so it can be inline with the preceeding number */
}
```
that you deleted in commit https://github.com/koreader/crengine/commit/aacb77c20d1adb9d3b28619ee66227671b926ab1#diff-4296cf0f47007e4df3fa206b9eb2dd87, then commit https://github.com/koreader/crengine/pull/384/commits/0b3a89967a99e2be1538bb7b860db65013f8dba2 does not change anything. In CoolReader this block not removed yet.
You wrote earlier 'I hope you won't fight much more with resurrecting legacy mode, when one can just switch to enhanced flat mode :)'.
So what is the best way to proceed? 'Fight much more with resurrecting legacy mode' (i.e. pick commit https://github.com/koreader/crengine/pull/384/commits/0b3a89967a99e2be1538bb7b860db65013f8dba2) or you just restore block
```css
body[name="comments"] section title p {
    display: inline; /* counteract default of block, so it can be inline with the preceeding number */
}
```
Or is this absolutely unnecessary block for enhanced rendering, so you removed it to simplify fb2.css?

virxkane commented:
The rest does not go beyond regressions.

poire-z commented:
> You wrote earlier 'I hope you won't fight much more with resurrecting legacy mode, when one can just switch to enhanced flat mode :)'.

Well, I spent about one hour finding this last fix, _before_ you called the fight off :) I'm glad the fight is off, but no reason to waste this fix, that looks like a logical followup to me.  

> So what is the best way to proceed? 'Fight much more with resurrecting legacy mode' (i.e. pick commit 0b3a899) or you just restore block
Or is this absolutely unnecessary block for enhanced rendering, so you removed it to simplify fb2.css?

Yes, I removed that bit of CSS because it wasn't needed anymore in enhanced mode.
I still think the 0b3a899 fix is more logical (it's totally related to what you added in that `if (legacy_rendering)` in `RenderFinalBlock()` which does a bit what's done elsewhere for enhanced mode) than fixing it via CSS.
(And we were fixing the limited handling of "display: run-in" generally - even if not used anywhere else - more than just fixing FB2 footnotes that the CSS would limit this too.)

virxkane commented:
Ok, thanks.
